include/envoy/upstream/cluster_manager.h

```diff
@@ -72,6 +72,69 @@ class ClusterUpdateCallbacksHandle {
 
 using ClusterUpdateCallbacksHandlePtr = std::unique_ptr<ClusterUpdateCallbacksHandle>;
 
+/**
+ * Status enum for the result of an attempted cluster discovery.
+ */
+enum class ClusterDiscoveryStatus {
+  /**
+   * The discovery process timed out. This means that there is no such cluster or we haven't yet
+   * received any reply from on-demand CDS about it.
+   */
+  Timeout,
+  /**
+   * Cluster found and currently available through ClusterManager.
+   */
+  Available,
+};
+
+/**
+ * ClusterDiscoveryCallback is a callback called at the end of the on-demand cluster discovery
+ * process. The status of the discovery is sent as a parameter.
+ */
+using ClusterDiscoveryCallback = std::function<void(ClusterDiscoveryStatus)>;
+using ClusterDiscoveryCallbackWeakPtr = std::weak_ptr<ClusterDiscoveryCallback>;
+using ClusterDiscoveryCallbackSharedPtr = std::shared_ptr<ClusterDiscoveryCallback>;
+
+/**
+ * ClusterDiscoveryCallbackHandle is a RAII wrapper for a ClusterDiscoveryCallback. Deleting the
+ * ClusterDiscoveryCallbackHandle will remove the callbacks from ClusterManager.
+ */
+class ClusterDiscoveryCallbackHandle {
+public:
+  virtual ~ClusterDiscoveryCallbackHandle() = default;
+};
+
+using ClusterDiscoveryCallbackHandlePtr = std::unique_ptr<ClusterDiscoveryCallbackHandle>;
+
+/**
+ * A handle to an on-demand CDS.
+ */
+class OdCdsApiHandle {
+public:
+  virtual ~OdCdsApiHandle() = default;
+
+  /**
+   * Request an on-demand discovery of a cluster with a passed name. This ODCDS may be used to
+   * perform the discovery process in the main thread if there is no discovery going on for this
+   * cluster. The passed callback will be invoked when the cluster is added and warmed up. It is
+   * expected that the callback will be destroyed when it is invoked. To cancel the discovery,
```

Comment:
What are the cross thread issues here? If the request is issued on a worker thread, will the passed callback be potentially invoked and destructed on the main thread, or do we guarantee this happens on the originating worker thread?

Reply:
The callback is always invoked in the same thread that requested the discovery. The worker thread is the sole owner of the callback, the main thread only has a weak access to it.

The expectation of the callback clearing itself is to avoid a situation where the callback could be called twice like:

Worker thread W1 requests the discovery of a cluster C. Main thread sees that this cluster is unknown and there is no discovery process for it ongoing, so it begins one. It sends a request, receives a reply and starts processing it in the main thread. Worker thread W2 in the same time also requests the discovery of cluster C (as it still not available for W2), so its callback is added to the discovery manager. In the main thread the processing of the reply is finished, so worker threads get notified about the added cluster C, so both W1 and W2 get notified (both had their callbacks in the discovery manager). And then in the main thread, the continuation of W2's discovery request is executed - it notices that the cluster C is already known, so it posts the callback back to W2 to be executed. And if the callback haven't cleared itself on the first call, it's going to be executed for the second time.

Please see the attached picture, hopefully makes the situation clearer:
![discovery-callbacks](https://user-images.githubusercontent.com/96081/115432222-6b61b100-a206-11eb-8e62-f4453485443a.png)

Reply:
This feels a bit fragile, why can't the discovery manager handle the clearing after invocation?

Regarding the diagram, this is excellent, can we include this somewhere? E.g. `source/docs`?

I think the thread model makes sense, can you explain at least the basic guarantee that callback is invoked on the thread it is registered on?

Reply:
Also, how are we managing lifetime? What happens if the original request that registered the callback is already deleted (including state captured in the callback) by time the request completes?

I'm asking as it's really tricky to get these corner cases correct, @stevenzzzz @chaoqin-li1123 @dmitri-d all have experience in implemented on-demand SRDS/VHDS and the tricky race conditions around lifetime. Do you have tests that capture all these possibilities?

Reply:
> This feels a bit fragile, why can't the discovery manager handle the clearing after invocation?
> 

This is a good idea, I'll investigate it. I can imagine `requestOnDemandClusterDiscovery` taking a `unique_ptr` to the callback, instead of `shared_ptr`, so the discovery manager becomes a sole manager of the callback. The returned handle would be a way to remove the callback from the manager. It should be a responsibility of a requester to make sure that whatever data are captured in the callback stay alive for the duration of the discovery or until the handle is destroyed.

I'll try to refactor it that way and see where it takes me.

> Regarding the diagram, this is excellent, can we include this somewhere? E.g. `source/docs`?
> 

Thanks. I can make some diagrams like this to explain the possible workflows during the discovery process. About this one specifically - it shows a scenario that we wanted to avoid, so it's probably not the best diagram for docs. :)

> I think the thread model makes sense, can you explain at least the basic guarantee that callback is invoked on the thread it is registered on?

Sure, will update the comment.

> Also, how are we managing lifetime? What happens if the original request that registered the callback is already deleted (including state captured in the callback) by time the request completes?

With current state of the code, both handle and the callback would also be deleted, so the discovery manager would either not see the callback (because destroying the handle should remove the callback) or will see it as a nullptr after locking (because it only has a weak_ptr to the callback), so it would not execute it.

> 
> I'm asking as it's really tricky to get these corner cases correct, @stevenzzzz @chaoqin-li1123 @dmitri-d all have experience in implemented on-demand SRDS/VHDS and the tricky race conditions around lifetime. Do you have tests that capture all these possibilities?

The tests capturing these possibilities are tricky to write as unit tests, because mock tls assumes just one thread, and mock dispatcher executes the callbacks immediately. Not sure if integration tests would make testing it easier either (no fine grained control of the discovery process and this feature is not really exposed yet in any way).

Reply:
@chaoqin-li1123 @dmitri-d  can you point to some example tests that you wrote that captured these lifetime issues for on-demand SRDS or VHDS?

Reply:
Also, if you are relying shared/weak pointer semantics for lifetime safety, don't worry about tryign to convert to a unique_ptr..

Reply:
I refactored it anyway, because I found some lifetime issues with handles (basically make a request for cluster foo, wait for the callback to be invoked, but keep the handle, request the same cluster again and then destroy the old handle, all of it in the same thread and we end up using a dangling iterator inside the destroyed handle).

Reply:
Here's a vhds integration test: https://github.com/envoyproxy/envoy/blob/main/test/integration/vhds_integration_test.cc#L668